### DISCUSSION TITLE
Add generic system to automatically create a record in a given dependent table

### DISCUSSION THREAD

**Discussion by grantfitzsimmons** [specify/specify7#5993] [Feature Requests]
**Requested by:** Philippe Verley at IRD on the [Speciforum](https://discourse.specifysoftware.org/t/automatically-create-a-new-collector-upon-ce-creation/2051/1)

> The title purposefully mimics [Automatically create a new Preparation, Determination, or Collection Object Attribute upon CO creation](https://discourse.specifysoftware.org/t/automatically-create-a-new-preparation-determination-or-collection-object-attribute-upon-co-creation/1014) and says it all.
> 
> The `Embedded Collecting Event` option and the three remote preferences `CO_CREATE_COA`, `CO_CREATE_DET` & `CO_CREATE_PREP` proved very useful for easing up and securing new CO entry. Number of clicks is decreased and mandatory fields are clearly identified. 
> 
>![image](https://github.com/user-attachments/assets/29736d66-9c90-4900-8263-2ee075e96b6f)
> 
> A `Collecting Event` does not make sense without a `Collector` in CAY herbarium. When I set up the `CO_CREATE` preps, my colleagues immediately asked whether a new `Collector` could be automatically created as well, since it is mandatory in our data entry process.
> 
> :point_up: **Can one automatically creates a new Collector upon CE creation ?**
> 
> A quick look at [`CollectionObject` initialization](https://github.com/specify/specify7/blob/dfe5887c006551322759ea5ada539036f5a68d57/specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts#L146-L157) and [`CollectionObject Resources` initialization](https://github.com/specify/specify7/blob/dfe5887c006551322759ea5ada539036f5a68d57/specifyweb/frontend/js_src/lib/components/DataModel/scoping.ts#L41-L75) seems to indicate that presently answer is no and that it may not be straightforward to implement.
> 
> :thinking: I'm also wondering why you decided to implement COA, DET & PRE creation through Remote Preferences ?  (*genuine question*) Have you thought of a **generalized approach** in `viewdef` resources, with for instance and XML attribute that would control the underlying resource creation ? 
> 
> ```xml
> 
> <cell type="subview" viewname="CollectionObjectAttribute" id="10" name="collectionobjectattribute" colspan="14" initialize="" create="onDataEntry"/>
> 
> <cell type="subview" viewname="Collector" id="12" name="collectors" colspan="12" rows="3" create="onDataEntry" />
> ```
> 
> The `create` attribute could accept several values:
> * `never` (default)
> * `onDataEntry` : create the underlying object on new data entry only
> * `ifEmpty` : create the underlying object on new data entry or on data edition if empty
> * ?
> 
> Feel free to split the last bit to a separate thread "Generalizing automatic object creation on Data Entry" if the thought interferes with my 1st question on collector creation.

**Comment by specifysoftware** [comment]
This issue has been mentioned on **Specify Community Forum**. There might be relevant details there:

https://discourse.specifysoftware.org/t/automatically-create-a-new-collector-upon-ce-creation/2051/2

**Comment by grantfitzsimmons** [comment]
From [Nate Shoobs on the Speciforum](https://discourse.specifysoftware.org/t/automatically-create-a-new-preparation-determination-or-collection-object-attribute-upon-co-creation/1014/6?u=grant):

> Quick question on this for anyone who knows – is there an analogous app resource definition for making a new preparation attribute record automatically for any new preparation?
> I’m imagining it’s a remote preference line that says
> “PREP_CREATE_[something]” but don’t know what the correct abbreviated form of Prep attribute would be for this.
> Is there a list somewhere of all of the xml-ified names for different tables?

**Comment by lexiclevenger** [comment]
From Alexis Beck at the Natural History Museum of Geneva via support ticket:

> When I create a new locality, I’d like that the locality detail associated is created at the same time so the user can visualize all the fields. Is it possible?
